### PR TITLE
Mob processing smoothing

### PR DESCRIPTION
**Pull request by TiviPlus**
## About The Pull Request

based on https://github.com/tgstation/TerraGov-Marine-Corps/pull/4305 but modified for tg

## Why It's Good For The Game

Less lagspikes every 2 seconds good

## Changelog
:cl:
code: Smoothed mob processing
/:cl:

### CODE REVIEW

**Review by nemvar — DISMISSED** [review]

File: code/controllers/subsystem/mobs.dm [hunk 1]
```diff
@@ -26,7 +37,24 @@ SUBSYSTEM_DEF(mobs)
 /datum/controller/subsystem/mobs/fire(resumed = 0)
 	var/seconds = wait * 0.1
 	if (!resumed)
-		src.currentrun = GLOB.mob_living_list.Copy()
+		if(crate == 1)
+			var/most = -1
+			var/least = 999999
```
Comment (nemvar) [inline code/controllers/subsystem/mobs.dm:42]: maybe this so it's immediatly clear that it's just some high number without concrete meaning.
```suggestion
			var/least = INFINITY
```

**Comment by optimumtact** [comment]
subsystems are already designed to stop if they're going over a tick, this should not be causing any lag spikes

@spookydonut what was your thinking around this.

@MrStonedOne

File: code/controllers/subsystem/mobs.dm [hunk 2]
```diff
@@ -40,3 +68,4 @@ SUBSYSTEM_DEF(mobs)
 			GLOB.mob_living_list.Remove(L)
 		if (MC_TICK_CHECK)
```
Comment (optimumtact) [inline code/controllers/subsystem/mobs.dm:69]: this is the line where they will yield if going over a tick

**Comment by spookydonut** [comment]
> 
> 
> subsystems are already designed to stop if they're going over a tick, this should not be causing any lag spikes
> 
> @spookydonut what was your thinking around this.
> 223% tick allocation just means the system is yeilding over multiple ticks, all you're doing is hiding this by splitting the runs up into buckets, which has the same outcome???
> @MrStonedOne

@optimumtact the idea was that heavy hitting subsystems beat sync sometimes to all process on the same tick, it wasn't just that mob tick usage was reduced, it also avoided the slight overhead from having to pause/resume.  It also gave more time on those 'super ticks' to other subsystems like lighting, in general I saw a slight decrease in overall tick usage %.  
The theory was that this could help avoid things causing tick overrun, which while I don't think has had a measurable impact on tgmc (mostly because we don't have LINDA) could help on tg. 

I think as a design goal we could aim to have things like machines, mobs, objects, etc that process regularly, tick slightly earlier/later if there's ticks significantly under-running.  This particular PR doesn't do that but I think it's a step towards some stonedmc improvements we could make.

**Comment by AnturK** [comment]
Don't we already stagger fires so multiple subsystems won't fire at the same time ? Unless they're in same blocked and resumed group that is.

**Comment by spookydonut** [comment]
> 
> 
> Don't we already stagger initial fires so multiple subsystems won't fire at the same time ? Unless they're in same blocked and resumed group that is.

@AnturK pretty sure stuff like lighting runs every tick

**Comment by optimumtact** [comment]
I'm not convinced this is doing anything.

If you're able to walk me through why this would do anything, or get mso to confirm it works, then I would change my mind.

**Pull request closed without merging**